Exercises.AsyncAwait/FileReader.cs

```diff
@@ -0,0 +1,192 @@
+        }
+
+
+        public async Task<bool> GetUrlsFromLocalFileAsync(string pathToFileWithUrls)
+        {
+            _logger.Write("Start reading file");
+
+            return await Task.Run(() =>
+            {
+                try
+                {
+                    _logger.Write($"started getting URLs from file {_urlsFileName}");
+
+                    foreach (string url in File.ReadLines(_urlsFileName))
+                    {
+                        results[url] = null;
+                    }
+                    _logger.Write("Getting URls from file successfully");
+
+                    return true;
+                }
+                catch (FileNotFoundException)
+                {
+                    _logger.Write("File not found");
+                }
+                catch (ArgumentException)
+                {
+                    _logger.Write("Path is a zero-length string");
+                }
+                catch (SecurityException)
+                {
+                    _logger.Write("You don't have the required permission");
+                }
+
+                return false;
+            });
+        }
+
+        public async void GetDataFromUrlOneByOneAsync()
+        {
+            decimal quantumPercent = (decimal)100/results.Count;
+            for (int i = 0; i < results.Count; i++)
+            {
+                var currentKeyValuePair = results.ElementAt(i);
+                results[currentKeyValuePair.Key] = await GetStringFromUrlAsync(currentKeyValuePair.Key);
+                _statusLine.Update(quantumPercent);
+            }
+        }
+
+        public async void GetDataFromUrlParallelAsync(int countDownloads)
+        {
+            decimal quantumPercent = (decimal)100 / results.Count;
+            if (countDownloads <= 0)
+            {
+                throw new ArgumentException("СountDownloads must be greater than zero");
+            }
+            else
+            {
+                var queueOfDownloads = new Queue<KeyValuePair<string, string>>(results);
+
+                while (queueOfDownloads.Count != 0)
+                {
+                    int countOfTasks = 0;
+                    if (queueOfDownloads.Count >= countDownloads)
+                    {
+                        countOfTasks = countDownloads;
+                    }
+                    else
+                    {
+                        countOfTasks = queueOfDownloads.Count;
+                    }
+                    var tasksArray = new Task<string>[countOfTasks];
+                    var urlsArray = new string[countOfTasks];
+
+                    for (int i = 0; i < countOfTasks; i++)
+                    {
+                        var currentDownloadUrl = (queueOfDownloads.Dequeue()).Key;
+                        urlsArray[i] = currentDownloadUrl;
+                        tasksArray[i] = GetContentFromUrlAsync(currentDownloadUrl);
+                    }
+
+                    Task.WaitAll(tasksArray);
```

Comment:
This kind of useless because it returns Task which is resolved once all tasks are completed. You should `await` for it